// @mysten/sui/dist/esm/transactions/plugins/utils.js
import { isValidNamedPackage, isValidNamedType } from "../../utils/move-registry.js";
const NAME_SEPARATOR = "@";
const findTransactionBlockNames = (builder) => {
  const packages = /* @__PURE__ */ new Set();
  const types = /* @__PURE__ */ new Set();
  for (const command of builder.commands) {
    if (!("MoveCall" in command)) continue;
    const tx = command.MoveCall;
    if (!tx) continue;
    const pkg = tx.package.split("::")[0];
    if (pkg.includes(NAME_SEPARATOR)) {
      if (!isValidNamedPackage(pkg)) throw new Error(`Invalid package name: ${pkg}`);
      packages.add(pkg);
    }
    for (const type of tx.typeArguments ?? []) {
      if (type.includes(NAME_SEPARATOR)) {
        if (!isValidNamedType(type)) throw new Error(`Invalid type with names: ${type}`);
        types.add(type);
      }
    }
  }
  return {
    packages: [...packages],
    types: [...types]
  };
};
const replaceNames = (builder, cache) => {
  for (const command of builder.commands) {
    const tx = command.MoveCall;
    if (!tx) continue;
    const nameParts = tx.package.split("::");
    const name = nameParts[0];
    if (name.includes(NAME_SEPARATOR) && !cache.packages[name])
      throw new Error(`No address found for package: ${name}`);
    nameParts[0] = cache.packages[name];
    tx.package = nameParts.join("::");
    const types = tx.typeArguments;
    if (!types) continue;
    for (let i = 0; i < types.length; i++) {
      if (!types[i].includes(NAME_SEPARATOR)) continue;
      if (!cache.types[types[i]]) throw new Error(`No resolution found for type: ${types[i]}`);
      types[i] = cache.types[types[i]];
    }
    tx.typeArguments = types;
  }
};
const listToRequests = (names, batchSize) => {
  const results = [];
  const uniqueNames = deduplicate(names.packages);
  const uniqueTypes = deduplicate(names.types);
  for (const [idx, name] of uniqueNames.entries()) {
    results.push({ id: idx, type: "package", name });
  }
  for (const [idx, type] of uniqueTypes.entries()) {
    results.push({
      id: idx + uniqueNames.length,
      type: "moveType",
      name: type
    });
  }
  return batch(results, batchSize);
};
const deduplicate = (arr) => [...new Set(arr)];
const batch = (arr, size) => {
  const batches = [];
  for (let i = 0; i < arr.length; i += size) {
    batches.push(arr.slice(i, i + size));
  }
  return batches;
};
export {
  findTransactionBlockNames,
  listToRequests,
  replaceNames
};
