// gql.tada/dist/gql-tada.mjs
import { Kind as r, parse as n } from "@0no-co/graphql.web";

var a = 0;

var e = new Set;

function initGraphQLTada() {
  function graphql(i, t) {
    var o = n(i).definitions;
    var s = new Set;
    for (var f of t || []) {
      for (var d of f.definitions) {
        if (d.kind === r.FRAGMENT_DEFINITION && !s.has(d)) {
          o.push(d);
          s.add(d);
        }
      }
    }
    var u;
    if ((u = o[0].kind === r.FRAGMENT_DEFINITION) && o[0].directives) {
      o[0].directives = o[0].directives.filter((r => "_unmask" !== r.name.value));
    }
    return {
      kind: r.DOCUMENT,
      definitions: o,
      get loc() {
        if (u) {
          var r = i + function concatLocSources(r) {
            try {
              a++;
              var n = "";
              for (var i of r) {
                if (!e.has(i)) {
                  e.add(i);
                  var {loc: t} = i;
                  if (t) {
                    n += t.source.body;
                  }
                }
              }
              return n;
            } finally {
              if (0 == --a) {
                e.clear();
              }
            }
          }(t || []);
          return {
            start: 0,
            end: r.length,
            source: {
              body: r,
              name: "GraphQLTada",
              locationOffset: {
                line: 1,
                column: 1
              }
            }
          };
        }
      }
    };
  }
  graphql.scalar = function scalar(r, n) {
    return n;
  };
  graphql.persisted = function persisted(n, a) {
    return {
      kind: r.DOCUMENT,
      definitions: a ? a.definitions : [],
      documentId: n
    };
  };
  return graphql;
}

function parse(r) {
  return n(r);
}

function readFragment(...r) {
  return 2 === r.length ? r[1] : r[0];
}

function maskFragments(r, n) {
  return n;
}

function unsafe_readResult(r, n) {
  return n;
}

var i = initGraphQLTada();

export { i as graphql, initGraphQLTada, maskFragments, parse, readFragment, unsafe_readResult };
